Configuration unification

ballPointPenguin opened the pull request:
TODO:

- [x] Consolidate all instances of `process.env` in server into the Config module.
- [x] Basic tests for Configuration and Environment concerns
- [x] remove all config .json files and replace with env
- [x] consolidate all env concerns on the top level
- [x] document every env variable and if/why/how to use it
- [ ] ~~Clarify what DEV_ MODE does and see if its impact can be reduced or split into multiple toggles~~
- [x] update README and CONFIGURATION files
  - [x] top-level README
  - [x] math README
  - [x] server README
  - [x] client-admin README
  - [x] client-participation README
  - [x] client-report README
  - [x] top-level CONFIGURATION
  - [x] math CONFIGURATION

## Top level ENV

_env` vars passed into client-apps via docker-compose:_

**Client Admin**
- FB_APP_ID

**Client Participation**
- EMBED_SERVICE_HOSTNAME
- FB_APP_ID

**Client Report**
- REPORT_UPLOADER default local
- REPORT_SERVICE_URL
- S3_BUCKET_PREPROD
- S3_BUCKET_PROD
- SCP_SUBDIR_PREPROD
- SCP_SUBDIR_PROD

**File Server itself**
- PORT (as STATIC_FILES_PORT) default 8080

### Personal Preferences

_fight me ;)_

1. Don’t rely on setting things in the users shell environment, e.g. with direnv or use of ‘exports’
1. Client apps should be buildable and debuggable without passing any non-default vars in on the command line (dev defaults are built-in)
1. Don’t rely on copying or writing files via scripts or shell commands
1. Avoid additional libraries or fancy tools as much as possible (convict)
1. Ultimately as long as it’s in the service’s env it should be good, and as such it is up to the admin of any particular deployment to manage the environments of their services as they see fit, regardless of if they use docker or not.
1. In development, `docker-compose up --build` should “just work”
1. Don’t rely on any manipulations via Makefile
1. Don’t duplicate env or config files in multiple places
1. For _server_ which has the lion's share of vars, prefer **dotenv** and/or just plain environment (set however) over complex yaml configurations
1. For _server_ isolate all calls to `process.env` into a single config.js module and use code (typescript) to ensure types, set defaults, etc. This is simpler and more flexible than convict and can be easily tested with jest unit test.

ballPointPenguin committed:
internal nginx ports must remain as 80/443

ballPointPenguin committed:
remove refrence to .env and .env_dev in server *ignore files;
slight update to *ignore files in server and admin.

ballPointPenguin committed:
remove unused env files and 'x' from server

ballPointPenguin committed:
template for example.env

ballPointPenguin committed:
update server README

ballPointPenguin committed:
configure basic jest api testing

ballPointPenguin committed:
touch up config module

ballPointPenguin committed:
clean up jest/test config and get tests to pass

ballPointPenguin committed:
fix test env

ballPointPenguin committed:
fix typo

ballPointPenguin committed:
DEV_MODE=true in test env

ballPointPenguin committed:
normalize Config properties

ballPointPenguin committed:
continuing to normalize env vars in server

ballPointPenguin committed:
remove hard-coded pol.is and localhost instances

ballPointPenguin committed:
refactor virtually all proess.env calls in server;

ballPointPenguin committed:
temporarily re-label original env for makefile

ballPointPenguin committed:
top-level example.env; gitignore .env

ballPointPenguin committed:
helpful note on server Dockerfile

ballPointPenguin committed:
create new example.env; remove old docker-dev.env

ballPointPenguin committed:
re-name and re-organize a few variables

ballPointPenguin committed:
remove/ignore top-level .env file

ballPointPenguin committed:
update example.env

ballPointPenguin committed:
use only the one .env for docker compose

ballPointPenguin committed:
remove unused config from math

ballPointPenguin committed:
remove unused docker-db-dev.env

ballPointPenguin committed:
minimize server test.env

ballPointPenguin committed:
typo and ordering fix in example.env

ballPointPenguin committed:
totally remove/replace all polis.config.js in client apps

ballPointPenguin committed:
clean up the url.js modules a bit

ballPointPenguin committed:
simplify development env in server

ballPointPenguin committed:
need to specify localhost for local server dev

ballPointPenguin committed:
pass MATH_ENV into math container

ballPointPenguin commented on server/src/server.ts:
```diff
@@ -228,11 +228,7 @@ akismet.verifyKey(function (err: any, verified: any) {
-// ^^^ possible to use localhost on Heroku?
-//  SELF_HOSTNAME = process.env.SERVICE_HOSTNAME
-//}
+// let SELF_HOSTNAME = Config.gerServerHostname();
```
Config.getServerHostname()

ballPointPenguin committed:
update README

ballPointPenguin committed:
update configuration.md

ballPointPenguin committed:
math and server docs

ballPointPenguin committed:
typo

ballPointPenguin committed:
remaining README updates

ballPointPenguin commented:
I will leave the removal of SCP concerns to another PR

ballPointPenguin committed:
remove extra asterisks from markdoow

xeeg commented:
I built this PR locally and login doesn't seem to work. I get "unauthorized domain: http://localhost" 
I believe this is the consequence of removing the whitelisting. 
https://github.com/compdemocracy/polis/issues/1153

xeeg commented:
I added localhost here to see if the whitelist was the only issue but signin still doesn't work. 

This is what I see in the log:

> polis-dev-server-1       | info errCompare, result null true
> polis-dev-server-1       | info startSession
> polis-dev-server-1       | info startSession: token set.
> polis-dev-server-1       | info uid 1
> polis-dev-server-1       | info lti_user_id undefined
> polis-dev-server-1       | info lti_context_id undefined

ballPointPenguin commented:
I clean an rebuild this branch from scratch with default config and see if I can find any problems with domain checking.

ballPointPenguin committed:
add a Quickstart and reference to make commands in README

ballPointPenguin committed:
update all instances of "docker-compose" to "docker compose"

ballPointPenguin committed:
include localhost:5000 in whitelist;
update API_DEV_HOSTNAME to localhost. This should allow
for more flexibility in DEV_MODE on either port 80 or 5000.

metasoarous reviewed (approved):
Looks great; Thanks!
